File: stage_zero.py
from typing import Any, Dict

import requests

from config import logger, Config
from utils import clean_url, handle_promotion


class StageZeroHandler:
    @classmethod
    def create_modal_view(cls, channel_id: str) -> dict:
        return {
            "type": "modal",
            "title": {
                "type": "plain_text",
                "text": "DevOps Stage 0",
            },
            "blocks": [
                {
                    "type": "section",
                    "text": {
                        "type": "mrkdwn",
                        "text": "Please enter the URL where you deployed your site.\n\nMake sure:\n• The URL is accessible\n• It contains 'Welcome to DevOps Stage 0'\n• It's served by NGINX",
                    },
                },
                {
                    "type": "input",
                    "block_id": "deployed_url",
                    "label": {"type": "plain_text", "text": "Deployed URL"},
                    "element": {
                        "type": "plain_text_input",
                        "action_id": "deployed_url",
                        "placeholder": {
                            "type": "plain_text",
                            "text": "Enter the full URL (including http:// or https://)",
                        },
                    },
                },
            ],
            "close": {"type": "plain_text", "text": "Cancel"},
            "submit": {"type": "plain_text", "text": "Submit"},
            "callback_id": "submission",
            "private_metadata": f"{channel_id}:0",
        }

    @classmethod
    def handle_submission(cls, user_id, channel_id, body, client):
        """Handle Stage 0 submission"""
        try:
            url = body["view"]["state"]["values"]["deployed_url"][
                "deployed_url"
            ]["value"]

            result = cls.grade(url, user_id)

            message = cls.format_response(result, user_id)
            if result.get("score", 0) == 3:
                try:
                    current = Config.STAGE_CHANNELS[0]["current"]
                    next_channels = Config.STAGE_CHANNELS[0]["next"]
                    handle_promotion(
                        client,
                        user_id,
                        current,
                        next_channels,
                        Config.STAGE_STATUS_EMOJIS[1],
                        Config.SLACK_USER_TOKEN,
                    )
                    new_channels = [
                        ch for ch in next_channels if ch not in current
                    ]
                    channels_text = ", ".join(
                        [f"<#{ch}>" for ch in new_channels]
                    )
                    client.chat_postMessage(
                        channel=user_id,
                        text=f"{message}\n\nYou should now be able to see the new channels: {channels_text} 🎉",
                    )
                except Exception as e:
                    logger.error(f"Promotion error: {str(e)}")
                    message += "\n\n⚠️ Promotion error. Please contact a mentor - <@U08A036CLGG>."
                    client.chat_postEphemeral(
                        channel=channel_id, user=user_id, text=message
                    )
            else:
                logger.debug(result)
                client.chat_postEphemeral(
                    channel=channel_id, user=user_id, text=message
                )

        except Exception as e:
            logger.error(f"Submission error: {str(e)}")
            client.chat_postEphemeral(
                channel=channel_id,
                user=user_id,
                text=":wrench: Something went wrong. Please try again.",
            )

    @classmethod
    def grade(cls, url: str, user_id: str) -> Dict[str, Any]:
        """Grade the Stage 0 submission"""
        clean_submission_url = clean_url(url)
        result = {
            "valid_url": False,
            "message_found": False,
            "is_nginx": False,
            "server_info": "Unknown",
            "score": 0,
        }

        if not clean_submission_url:
            return result

        try:
            resp = requests.get(clean_submission_url, timeout=10)
            resp.raise_for_status()

            result["valid_url"] = True
            result["score"] += 1

            result["message_found"] = "Welcome to DevOps Stage 0" in resp.text
            if result["message_found"]:
                result["score"] += 1

            server_info = resp.headers.get("Server", "Unknown")
            result["server_info"] = server_info
            result["is_nginx"] = "nginx" in server_info.lower()
            if result["is_nginx"]:
                result["score"] += 1

        except requests.exceptions.RequestException:
            pass

        return result

    @classmethod
    def format_response(cls, result: Dict[str, Any], user_id: str) -> str:
        """Format the grading results"""
        msg = [
            f"Hey <@{user_id}>! Here's your Stage 0 result:",
            "",
            "*Requirements:*",
            f"• *Valid URL:* {':white_check_mark: Passed' if result['valid_url'] else ':x: Failed'}",
            f"• *Welcome Message:* {':white_check_mark: Passed' if result['message_found'] else ':x: Failed'}",
            f"• *NGINX Server:* {':white_check_mark: Passed' if result['is_nginx'] else ':x: Failed'}",
            "",
            f"*Server Detected:* {result['server_info']}",
            f"*Score: {result['score']}/3*",
        ]

        if result["score"] < 3:
            msg.extend(["", "*To improve your score:*"])
            if not result["valid_url"]:
                msg.append("• Ensure your URL is properly formatted")
            if not result["message_found"]:
                msg.append("• Add 'Welcome to DevOps Stage 0' to your page")
            if not result["is_nginx"]:
                msg.append("• Configure your server to use NGINX")
            msg.extend(
                ["", "Make these changes and use *`/submit`* to try again!"]
            )

        return "\n".join(msg)

File: src/stages/stage-zero.ts
import fetch from 'node-fetch';
import { BaseStage } from './base-stage';
import { GoogleSheetService } from '@/services/google-sheets';
import { logger, Config } from '@/config';
import { handlePromotion, checkUrlUniqueness, extractSubmissionData } from '@/utils';
import type { StageResult, SlackModalView, SlackSubmissionBody, SubmissionData } from '@/types';

interface StageZeroResult {
  readonly deployedValid: boolean;
  readonly messagePresent: boolean;
  readonly nginxPresent: boolean;
  readonly blogValid: boolean;
  readonly backlinkPresent: boolean;
  readonly server: string;
  readonly errors: readonly string[];
}

export class StageZero extends BaseStage {
  readonly emoji = ':zero:';
  readonly channels = ['C089GSHEMFT'] as const;
  readonly nextChannels = ['C089GSHEMFT', 'C08AHHWBTK8', 'C08B3UKM0QN'] as const;
  readonly requiredScore = 5;
  readonly expectedText = 'Welcome to DevOps Stage 0';

  readonly backlinks = [
    'https://hng.tech/hire/devops-engineers',
    'https://hng.tech/hire/cloud-engineers',
    'https://hng.tech/hire/site-reliability-engineers',
    'https://hng.tech/hire/platform-engineers',
    'https://hng.tech/hire/infrastructure-engineers',
    'https://hng.tech/hire/kubernetes-specialists',
    'https://hng.tech/hire/aws-solutions-architects',
    'https://hng.tech/hire/azure-devops-engineers',
    'https://hng.tech/hire/google-cloud-engineers',
    'https://hng.tech/hire/ci-cd-pipeline-engineers',
    'https://hng.tech/hire/monitoring-observability-engineers',
    'https://hng.tech/hire/automation-engineers',
    'https://hng.tech/hire/docker-specialists',
    'https://hng.tech/hire/linux-developers',
    'https://hng.tech/hire/postgresql-developers',
  ] as const;

  protected readonly sheet = new GoogleSheetService({
    spreadsheetId: '1t-JU71GkCOlYf7nAWdJWxoJpzNcDiKlqfB4aZDmDiAg',
    columns: {
      A: 'timestamp',
      B: 'username',
      C: 'user_id',
      D: 'trials',
      E: 'deployed_url',
      F: 'blog_url',
      G: 'promoted',
    },
  });

  submissionView(channelId: string): SlackModalView {
    return {
      type: 'modal',
      title: { type: 'plain_text', text: 'DevOps Stage 0' },
      blocks: [
        {
          type: 'input',
          block_id: 'deployed_url',
          label: { type: 'plain_text', text: 'Deployed URL' },
          element: {
            type: 'plain_text_input',
            action_id: 'deployed_url',
            placeholder: {
              type: 'plain_text',
              text: 'Enter the full URL (including http:// or https://)',
            },
          },
        },
        {
          type: 'input',
          block_id: 'blog_url',
          label: { type: 'plain_text', text: 'Blog Post URL' },
          element: {
            type: 'plain_text_input',
            action_id: 'blog_url',
            placeholder: {
              type: 'plain_text',
              text: 'Enter your blog post URL',
            },
          },
        },
      ],
      close: { type: 'plain_text', text: 'Cancel' },
      submit: { type: 'plain_text', text: 'Submit' },
      callback_id: 'submission',
      private_metadata: channelId,
    };
  }

  async submit(
    channelId: string,
    body: SlackSubmissionBody,
    client: unknown // Using unknown for now due to complex Slack client typing
  ): Promise<void> {
    try {
      const submissionData = extractSubmissionData(body);
      const { userId, username } = submissionData;

      // Check if already promoted
      if (await this.checkIfAlreadyPromoted(userId)) {
        await (client as any).chat.postEphemeral({
          channel: channelId,
          user: userId,
          text: '🎉 You have already passed Stage 0! No need to submit again.',
        });
        return;
      }

      // Extract URLs from submission
      const values = body.view.state.values;
      const deployedUrl = values.deployed_url?.deployed_url?.value;
      const blogUrl = values.blog_url?.blog_url?.value;

      if (!deployedUrl || !blogUrl) {
        await (client as any).chat.postEphemeral({
          channel: channelId,
          user: userId,
          text: '❌ Both deployed URL and blog URL are required.',
        });
        return;
      }

      // Validate URL uniqueness
      for (const [url, urlType] of [
        [deployedUrl, 'deployed_url'],
        [blogUrl, 'blog_url'],
      ] as const) {
        const { isUnique, errorMessage } = await checkUrlUniqueness(
          this.sheet,
          url,
          userId,
          urlType
        );

        if (!isUnique) {
          await (client as any).chat.postEphemeral({
            channel: channelId,
            user: userId,
            text: `❌ ${errorMessage}`,
          });
          return;
        }
      }

      // Grade submission
      const result = await this.processSubmission({
        ...submissionData,
        values: { deployedUrl, blogUrl },
      });

      const trials = await this.getTrialCount(userId);
      const promoted = result.passed;

      // Update submission record
      await this.updateSubmissionRecord(
        userId,
        username,
        {
          deployed_url: deployedUrl,
          blog_url: blogUrl,
        },
        promoted,
        trials
      );

      // Generate result message
      const message = this.formatResultMessage(result, userId, trials);

      if (promoted) {
        // Handle promotion
        await handlePromotion(
          client as any,
          userId,
          this.channels,
          this.nextChannels,
          this.emoji,
          Config.slack.userToken
        );

        const newChannels = this.nextChannels
          .filter(ch => !(this.channels as readonly string[]).includes(ch))
          .map(ch => `<#${ch}>`)
          .join(', ');

        await (client as any).chat.postMessage({
          channel: userId,
          text: `${message}\n\n🚀 Access granted to: ${newChannels}`,
        });
      } else {
        await (client as any).chat.postEphemeral({
          channel: channelId,
          user: userId,
          text: message,
        });
      }
    } catch (error) {
      logger.error('Stage 0 submission error:', error);
      await (client as any).chat.postEphemeral({
        channel: channelId,
        user: body.user.id,
        text: '🚨 An error occurred while processing your submission. Please try again or contact support.',
      });
    }
  }

  async processSubmission(data: SubmissionData): Promise<StageResult> {
    return this.gradeSubmission(data);
  }

  protected async gradeSubmission(data: SubmissionData): Promise<StageResult> {
    const { deployedUrl, blogUrl } = data.values as { deployedUrl: string; blogUrl: string };

    let score = 0;
    let details: StageZeroResult = {
      deployedValid: false,
      messagePresent: false,
      nginxPresent: false,
      blogValid: false,
      backlinkPresent: false,
      server: 'Unknown',
      errors: [],
    };

    const errors: string[] = [];

    // Check deployed URL
    const deployedResult = await this.fetchUrlContent(deployedUrl);
    if (deployedResult.success) {
      details = { ...details, deployedValid: true };
      score += 1;

      if (deployedResult.content.includes(this.expectedText)) {
        details = { ...details, messagePresent: true };
        score += 1;
      }

      const server = deployedResult.response?.headers?.get('server') ?? 'Unknown';
      details = { ...details, server };

      if (server.toLowerCase().includes('nginx')) {
        details = { ...details, nginxPresent: true };
        score += 1;
      }
    } else {
      errors.push(`Deployed URL: ${deployedResult.error}`);
    }

    // Check blog URL
    const blogResult = await this.fetchUrlContent(blogUrl);
    if (blogResult.success) {
      details = { ...details, blogValid: true };
      score += 1;

      if (this.checkBacklinks(blogResult.content)) {
        details = { ...details, backlinkPresent: true };
        score += 1;
      }
    } else {
      errors.push(`Blog URL: ${blogResult.error}`);
    }

    return {
      score,
      passed: score >= this.requiredScore,
      details: { ...details, errors },
      errors,
    };
  }

  private async fetchUrlContent(
    url: string,
    timeout: number = 15000
  ): Promise<{
    success: boolean;
    content: string;
    error?: string;
    response?: import('node-fetch').Response;
  }> {
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), timeout);

      const response = await fetch(url, {
        signal: controller.signal,
        redirect: 'follow',
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        return {
          success: false,
          content: '',
          error: `HTTP ${response.status}: ${response.statusText}`,
        };
      }

      const content = await response.text();
      return {
        success: true,
        content,
        response,
      };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      return {
        success: false,
        content: '',
        error: `Error accessing URL: ${errorMessage}`,
      };
    }
  }

  private checkBacklinks(content: string): boolean {
    return this.backlinks.some(backlink => content.includes(backlink));
  }

  protected formatResultMessage(result: StageResult, userId: string, trials: number): string {
    const details = result.details as unknown as StageZeroResult;
    const lines = [
      `<@${userId}> Stage 0 Results (Attempt #${trials}):\n`,
      '📋 Requirements Check:',
      `${details.deployedValid ? '✅' : '❌'} Deployed URL is accessible`,
      `${details.messagePresent ? '✅' : '❌'} Welcome message present`,
      `${details.nginxPresent ? '✅' : '❌'} NGINX server detected`,
      `${details.blogValid ? '✅' : '❌'} Blog post is accessible`,
      `${details.backlinkPresent ? '✅' : '❌'} Backlinks present`,
      `Server: ${details.server}`,
      `Score: ${result.score}/${this.requiredScore}\n`,
    ];

    if (details.errors.length > 0) {
      lines.push('⚠️ Errors encountered:');
      lines.push(...details.errors.map(error => `• ${error}`));
      lines.push('');
    }

    if (!result.passed) {
      lines.push('📝 Required improvements:');
      if (!details.deployedValid) {
        lines.push('• Ensure your deployed URL is accessible');
      }
      if (!details.messagePresent) {
        lines.push(`• Add '${this.expectedText}' to your page`);
      }
      if (!details.nginxPresent) {
        lines.push('• Configure NGINX as your web server');
      }
      if (!details.blogValid) {
        lines.push('• Ensure your blog post URL is accessible');
      }
      if (!details.backlinkPresent) {
        lines.push('• Include at least one of the provided backlinks in your blog post');
      }
      lines.push("\n💡 Resubmit when you've made these improvements!");
    } else {
      lines.push(
        `🎉 Congratulations! You've completed Stage 0 in ${trials} ${
          trials === 1 ? 'attempt' : 'attempts'
        }!`
      );
    }

    return lines.join('\n');
  }
}
